### PR TITLE
Closure sizes

### PR DESCRIPTION
**Pull request by gilligan**
This PR introduces changes to reduce the closure size of our deployments:

1. **buildPursProject**: npm dependencies are split into production and dev dependencies. The latter can be removed when deploying. This PR runs `npm prune` to reduce the `node_modules` folder to production dependencies. The playgrounds don't actually need the `node_modules` at all and the change removes the `node_modules` from the closure completely.
2. **runtimeGhc closure**: The closure of the `server-invoker` is gigantic. This PR changes the packages added via `ghcWithPackages` somewhat which reduces the scope.
3. **haddock data**: The plutus-playground tutorial contained the _complete_ haddock site data even though it isn't viewable (and not supposed to be viewable) as part of the tutorial. This makes for a difference of 800mb+ in the closure

**Notes**:
- Unfortunately i don't have figures available right now but there was a 1GB+ different in the closure size for initial deployments of the playground configurations.
- For testing I started the playgrounds locally (`npm start`) and then ran the `server-invoker` and tried compiling all available examples

----
Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
    - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
    - [ ] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues
- If you changed any Purescript files:
    - [ ] `nix-shell shell.nix --run fix-purty` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by michaelpj** [comment]
I thought the haddock *was* accessible from the Playground, and linked to from the tutorial?

File: nix/lib/purescript.nix [hunk 1]
```diff
@@ -49,4 +50,8 @@ stdenv.mkDerivation {
   installPhase = ''
     mv dist $out
   '';
+
+  postInstall = ''
+    npm prune --production
```
Comment (michaelpj) [inline nix/lib/purescript.nix:55]: comment?

File: nix/docs.nix [hunk 2]
```diff
@@ -66,6 +66,13 @@ pkgs.recurseIntoAttrs {
   plutus-report = import ../notes/plutus-report { inherit buildLatexDoc; };
   cost-model-notes = import ../notes/cost-model-notes { inherit buildLatexDoc; };
 
+  plutus-tutorial = pkgs.runCommand "plutus-tutorial" { } ''
```
Comment (michaelpj) [inline nix/docs.nix:69]: comment?

**Comment by gilligan** [comment]
> I thought the haddock was accessible from the Playground, and linked to from the tutorial?

I couldn't (easily) find a way to access it. From skimming over the tutorial there was no clear link to it either.

Considering the fact that the tutorial iltself is just a few MBs and haddock is like a GB it would actually nice if it does NOT have to be part of it :sweat_smile:  -- @j-mueller  ?

**Comment by silky** [comment]
unless i'm misunderstanding @gilligan it's linked at the top of the page right next to the "tutorial"; it's the "API" link:

image

i believe these changes will definitely impact https://github.com/input-output-hk/plutus/pull/2858 but i can fix that after the fact

i did think we _did_ want the haddocks available though?

**Comment by gilligan** [comment]
@silky oh wow - how did I manage to miss that. Well in that case I suppose I will have to add back the haddock stuff but maybe rephrase my question:

@michaelpj @j-mueller The haddock weighs in with a whooping 845MB closure size ;( It doesn't only provide docs for plutus/marlowe internal packages but plenty of external packages also - I suppose we could limit it to Plutus/Marlowe packages?

File: nix/lib/purescript.nix (hunk 1 above)
Comment (gilligan) [reply nix/lib/purescript.nix:64]: Added a comment explaining what is going on and why it is where it is

**Comment by j-mueller** [comment]
We don't have any other location where the haddocks are hosted publicly, so I would like to keep them in the Playground.

Ideally we would link to hackage haddocks for all non-iohk packages, but this seems fiddly. Failing that, removing all those other packages from our haddocs is OK.

I am a bit surprised at the size though!

**Comment by gilligan** [comment]
https://gist.github.com/gilligan/4b15a3022f8186a5adf0b82ae8e39d97
821MB in total.

I am pretty sure it would be much smaller if it was only for _our_ packages. The haddocks stuff is created in https://github.com/input-output-hk/plutus/blob/closure-sizes/nix/lib/haddock-combine.nix. Looking at `nix/pkgs/default.nix` i see:

```
  # combined haddock documentation for all public plutus libraries
  plutus-haddock-combined =
```

which appears to be misleading because it does actually include all dependencies. From what I understand @silky is currently working on that. Maybe there is a way to fix that? I have *no idea* at all if this would be straight forward or super complicated.

**Review by j-mueller — APPROVED** [review]
LGTM
